### PR TITLE
Add Key Event telemetry

### PR DESCRIPTION
**Pull request by yongyan-gh**

### CODE REVIEW

File: src/Sarif.Viewer.VisualStudio.Core/Telemetry/KeyEventTelemetry.cs [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+                {
+                    return string.Empty;
+                }
+
+                vsVersion ??= VsUtilities.GetVsVersion() ?? string.Empty;
+                return vsVersion;
+            }
+        }
+
+        public static string ExtensionVersion
```
Comment (EasyRhinoMSFT) [inline src/Sarif.Viewer.VisualStudio.Core/Telemetry/KeyEventTelemetry.cs:37]: >ExtensionVersion [](http://example.com/codeflow?start=29&length=16)

Is it possible to use the actual extension version here? Might be easier to associate with events

File: src/Sarif.Viewer.VisualStudio.Core/Telemetry/KeyEventTelemetry.cs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            {
+                throw new ArgumentNullException(nameof(eventName));
+            }
+
+            UserTaskEvent trackEvent = CreateEvent(eventName, result);
+
+            this.telemetryClient.PostEvent(trackEvent);
+        }
+
+        public void TrackEvent(string eventName, Dictionary<string, string> properties, TelemetryResult result = TelemetryResult.Success)
```
Comment (EasyRhinoMSFT) [inline src/Sarif.Viewer.VisualStudio.Core/Telemetry/KeyEventTelemetry.cs:63]: >Dictionary<string, string> properties, [](http://example.com/codeflow?start=49&length=38)

Make this the third parameter

File: src/Sarif.Viewer.VisualStudio.Core/Telemetry/ITelemetryProvider.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.VisualStudio.Telemetry;
+
+namespace Microsoft.Sarif.Viewer.Telemetry
+{
```
Comment (EasyRhinoMSFT) [inline src/Sarif.Viewer.VisualStudio.Core/Telemetry/ITelemetryProvider.cs:7]: Please add doc comments along with <inheritdoc .../> in the implementation class

File: src/Sarif.Viewer.VisualStudio.Core/Telemetry/KeyEventTelemetry.cs [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Reflection;
+
+using Microsoft.Sarif.Viewer.Shell;
+using Microsoft.VisualStudio.Telemetry;
+
+namespace Microsoft.Sarif.Viewer.Telemetry
+{
```
Comment (EasyRhinoMSFT) [inline src/Sarif.Viewer.VisualStudio.Core/Telemetry/KeyEventTelemetry.cs:13]: Please add doc comments for the class and all public methods

File: src/Sarif.Viewer.VisualStudio.Core/Telemetry/KeyEventTelemetry.cs [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+            UserTaskEvent trackEvent = CreateEvent(eventName, result);
+
+            foreach (KeyValuePair<string, string> propertie in properties)
+            {
+                if (!trackEvent.Properties.ContainsKey(propertie.Key))
+                {
+                    trackEvent.Properties[propertie.Key] = propertie.Value;
+                }
+            }
```
Comment (EasyRhinoMSFT) [inline src/Sarif.Viewer.VisualStudio.Core/Telemetry/KeyEventTelemetry.cs:83]: DRY
You can make a helper method for this.
void MergeEventProperties(TelemetryEvent instance, Dictionary<string, string> properties)

File: src/Sarif.Viewer.VisualStudio.Core/Services/DataService.cs [hunk 6]
```diff
@@ -108,6 +112,8 @@ private async System.Threading.Tasks.Task<int> SendEnhancedResultDataAsync(Sarif
                 this.sarifErrorListEventSelectionService.SelectedItem = items[0];
 
                 items[0].Locations?.FirstOrDefault()?.NavigateTo(usePreviewPane: false, moveFocusToCaretLocation: true);
+
+                this.keyEventTelemetry.TrackEvent("DisplayKeyEventData");
```
Comment (EasyRhinoMSFT) [inline src/Sarif.Viewer.VisualStudio.Core/Services/DataService.cs:116]: >DisplayKeyEventData [](http://example.com/codeflow?start=51&length=19)

Should we put these in some central location?

File: src/Sarif.Viewer.VisualStudio.Core/Telemetry/KeyEventTelemetry.cs (hunk 1 above)
Comment (yongyan-gh) [reply src/Sarif.Viewer.VisualStudio.Core/Telemetry/KeyEventTelemetry.cs:59]: as suggested, read the actual version from extension manifest file

File: src/Sarif.Viewer.VisualStudio.Core/Telemetry/KeyEventTelemetry.cs (hunk 2 above)
Comment (yongyan-gh) [reply src/Sarif.Viewer.VisualStudio.Core/Telemetry/KeyEventTelemetry.cs]: combined 2 functions and make the properties 3rd parameter.

File: src/Sarif.Viewer.VisualStudio.Core/Telemetry/ITelemetryProvider.cs (hunk 3 above)
Comment (yongyan-gh) [reply src/Sarif.Viewer.VisualStudio.Core/Telemetry/ITelemetryProvider.cs:7]: add comments

File: src/Sarif.Viewer.VisualStudio.Core/Telemetry/KeyEventTelemetryClient.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.VisualStudio.Telemetry;
+
+namespace Microsoft.Sarif.Viewer.Telemetry
+{
```
[the remark this answers is not in the document]
Comment (yongyan-gh) [reply src/Sarif.Viewer.VisualStudio.Core/Telemetry/KeyEventTelemetryClient.cs:7]: added

File: src/Sarif.Viewer.VisualStudio.Core/Telemetry/KeyEventTelemetry.cs (hunk 4 above)
Comment (yongyan-gh) [reply src/Sarif.Viewer.VisualStudio.Core/Telemetry/KeyEventTelemetry.cs:13]: added

File: src/Sarif.Viewer.VisualStudio.Core/Services/DataService.cs (hunk 6 above)
Comment (yongyan-gh) [reply src/Sarif.Viewer.VisualStudio.Core/Services/DataService.cs]: updated

File: src/Sarif.Viewer.VisualStudio.Core/Telemetry/KeyEventTelemetry.cs (hunk 5 above)
Comment (yongyan-gh) [reply src/Sarif.Viewer.VisualStudio.Core/Telemetry/KeyEventTelemetry.cs:96]: done

**Review by EasyRhinoMSFT — APPROVED** [review]
:shipit:

**Pull request merged**